src/dmd/expressionsem.d

```diff
@@ -8826,12 +8826,41 @@ private extern (C++) final class ExpressionSemanticVisitor : Visitor
 
                 // May be block or element-wise assignment, so
                 // convert e1 to e1[]
-                // flatten into a 1D array unless e2 is an array type
-                if (exp.op != TOK.assign && e2x.type.ty != Tarray && e2x.type.ty != Tsarray)
```

Comment:
This fix removes the assignment restriction and you have added test cases for it which is good, but was there a reason behind the restriction?

Reply:
This change was introduced in #3467 but i don't understand the reasoning to avoid assignment..

Reply:
Probably to facilitate the distinction regarding which types are allowed for construction and assignments.
Restored it anyway as the discussion whether this behaviour is desirable belongs in another PR